src/Simulation/qdk_sim_rs/src/instrument.rs

```diff
@@ -0,0 +1,153 @@
+                        (
+                            if result { 1 } else { 0 },
+                            State {
+                                n_qubits: state.n_qubits,
+                                data: StateData::Stabilizer(new_tableau),
+                            },
+                        )
+                    }
+                }
+            }
+        }
+    }
+
+    // TODO: Add more methods for making new instruments in convenient ways.
+
+    /// Returns a serialization of this instrument as a JSON object.
+    pub fn as_json(&self) -> String {
+        serde_json::to_string(&self).unwrap()
+    }
+}
+
+fn sample_effects(effects: &[Process], idx_qubits: &[usize], state: &State) -> (usize, State) {
+    let mut possible_outcomes = effects
+        .iter()
+        .enumerate()
+        .map(|(idx, effect)| {
+            let output_state = effect.apply_to(idx_qubits, state).unwrap();
+            let tr = (&output_state).trace();
+            (idx, output_state, tr.norm())
+        })
+        .collect::<Vec<_>>();
+    // TODO[perf]: Downgrade this to a debug_assert!, and configure the CI
+    //             build to enable debug_assertions at full_validation.
+    assert!(
+        possible_outcomes.iter().any(|post_state| post_state.1.trace().norm() >= 1e-10),
+        "Expected output of applying instrument to be nonzero trace.\nInstrument effects:\n{:?}\n\nInput state:\n{}\n\nPostselected states:\n{:?}",
+        effects, state, possible_outcomes
+    );
+    let mut rng = rand::thread_rng();
+    let random_sample: f64 = rng.gen();
+    for (idx, cum_pr) in possible_outcomes
+        .iter()
+        .scan(0.0f64, |acc, (_idx, _, pr)| {
+            *acc += *pr;
+            Some(*acc)
+        })
+        .enumerate()
+    {
+        if random_sample < cum_pr {
+            // In order to not have to copy the output state, we need
+            // to be able to move it out from the vector. To do so,
+            // we retain only the element of the vector whose index
+            // is the one we want and then pop it, leaving an empty
+            // vector (that is, a vector that owns no data).
+            possible_outcomes.retain(|(i, _, _)| idx == *i);
+            let (_, mut output_state, tr) = possible_outcomes.pop().unwrap();
+            if tr.abs() >= 1e-10 {
+                if let Mixed(ref rho) = output_state.data {
+                    output_state.data = Mixed(rho * (1.0f64 / tr));
+                } else {
+                    panic!("Couldn't renormalize, expected mixed output from instrument.");
+                }
+            }
+            assert!(
+                (output_state.trace() - 1.0).norm() <= 1e-10,
+                "Expected output of instrument to be trace 1."
+            );
+            return (idx, output_state);
+        }
+    }
+    let (idx, output_state, _) = possible_outcomes.pop().unwrap();
+    drop(possible_outcomes);
```

Comment:
Out of curiosity, why the explicit drop of `possible_outcomes` here? Wouldn't it automatically be dropped when it goes out of scope during the return below? of note, it does not get explicitly dropped before the early return in the if-statement above (line 147).

Reply:
Good catch...! I think this was needed back when I wrote the innovation week project back in December, as I was handling the return value a bit differently such that possible_outcomes holding owned values of other states was a problem here. I think refactoring since has obviated the need here, such that I can now drop this statement. Thanks for pointing this out!